File: dart/servewright_flutter/lib/src/transport.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:servewright_flutter/src/transition.dart';

abstract class ServewrightTransport {
  StreamSubscription<String>? connect(
    String screen, {
    required void Function(ServewrightTransition transition) onTransition,
    void Function(Object error)? onError,
  });
}

class SseTransport implements ServewrightTransport {
  SseTransport({this.streamUrl = '/servewright/stream', http.Client? client})
      : _client = client ?? http.Client();

  final String streamUrl;
  final http.Client _client;

  @override
  StreamSubscription<String>? connect(
    String screen, {
    required void Function(ServewrightTransition transition) onTransition,
    void Function(Object error)? onError,
  }) {
    final controller = StreamController<String>();
    final request = http.Request('GET', Uri.parse('$streamUrl/$screen'));
    request.headers['Accept'] = 'text/event-stream';

    _client.send(request).then((response) {
      response.stream
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .listen(
            controller.add,
            onError: onError,
          );
    }).catchError(onError ?? (_) {});

    String? dataBuffer;
    return controller.stream.listen((line) {
      if (line.startsWith('data:')) {
        dataBuffer = line.substring(5).trim();
      } else if (line.isEmpty && dataBuffer != null) {
        final transition = ServewrightTransition.fromJson(
          jsonDecode(dataBuffer!) as Map<String, dynamic>,
        );
        onTransition(transition);
        dataBuffer = null;
      }
    }, onError: onError);
  }
}

class ImmediateTransport implements ServewrightTransport {
  ImmediateTransport(this.transitions);

  final List<ServewrightTransition> transitions;

  @override
  StreamSubscription<String>? connect(
    String screen, {
    required void Function(ServewrightTransition transition) onTransition,
    void Function(Object error)? onError,
  }) {
    for (final transition in transitions) {
      onTransition(transition);
    }
    return null;
  }
}

File: js/servewright-react/src/transport.ts
import type { Transition } from "./types.js";

export interface TransportHandlers {
  onTransition: (transition: Transition) => void;
  onError?: (error: unknown) => void;
}

export interface Transport {
  connect(screen: string, handlers: TransportHandlers): () => void;
}

export class SseTransport implements Transport {
  constructor(private readonly streamUrl = "/servewright/stream") {}

  connect(screen: string, handlers: TransportHandlers): () => void {
    const source = new EventSource(`${this.streamUrl}/${encodeURIComponent(screen)}`);

    source.addEventListener("transition", (event) => {
      try {
        const transition = JSON.parse((event as MessageEvent<string>).data) as Transition;
        handlers.onTransition(transition);
      } catch (error) {
        handlers.onError?.(error);
      }
    });

    source.onerror = (error) => {
      handlers.onError?.(error);
    };

    return () => source.close();
  }
}

export class ImmediateTransport implements Transport {
  constructor(private readonly transitions: Transition[] = []) {}

  connect(_screen: string, handlers: TransportHandlers): () => void {
    for (const transition of this.transitions) {
      handlers.onTransition(transition);
    }
    return () => undefined;
  }
}
